lib/labels/labels.go

```diff
@@ -166,3 +169,118 @@ func (l *Dynamic) setLabel(name string, value types.CommandLabel) {
 
 	l.c.Labels[name] = value
 }
+
+// EC2LabelConfig is the configuration for the EC2 label service.
+type EC2LabelConfig struct {
+	Client utils.InstanceMetadata
+	Clock  clockwork.Clock
+	Log    *logrus.Entry
+}
+
+func (conf *EC2LabelConfig) checkAndSetDefaults() error {
+	if conf.Client == nil {
+		client, err := utils.NewInstanceMetadataClient(context.TODO())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		conf.Client = client
+	}
+	if conf.Clock == nil {
+		conf.Clock = clockwork.NewRealClock()
+	}
+	if conf.Log == nil {
+		conf.Log = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return nil
+}
+
+// EC2Labels is a service that periodically imports tags from EC2 via instance
+// metadata.
+type EC2Labels struct {
+	c      *EC2LabelConfig
+	mu     sync.RWMutex
+	once   sync.Once
+	labels map[string]string
+
+	closeContext context.Context
+	closeFunc    context.CancelFunc
+}
+
+func NewEC2Labels(ctx context.Context, c *EC2LabelConfig) (*EC2Labels, error) {
```

Comment:
Passing a context makes sense for an operation that may take a while (network call, for example).

In this case, `NewEC2Labels` always returns immediately - it doesn't need to support timeouts or cancelation.

I think it would be better for `*EC2Labels` to have a `Stop()` or `Close()` method, similar to a `time.Ticker` or a `net.Listener`.


Edit: I see you already have a `Close` method, which makes this extra confusing. Is the proper way to stop this service to cancel the context, or to call Close, or both?

Let's just have our own close channel that we write to or close when Close is called.